ISSUE-338 initial commit for invitation management

sbryzak opened the pull request:
This pull request implements the create invitation endpoint and invitation service.  The scope of the invitations API has been expanded since starting work on this issue to allow for creating invitations for users to accept a role for a resource, this functionality will be covered by a different issue.

Request to invite a user to an organization/team/security group:

```
POST /invitations/<identityID>
or
POST /invitations/<resourceID>

{  "data": [
  {  "identity-id": "aaaa-bbbb-cccc-dddd-1234-5678",
     "member": "true"},
  {  "identity-id": "aaaa-bbbb-cccc-dddd-1234-9999",
      "member": "true"},
  {  "identity-id": "aaaa-bbbb-cccc-dddd-1234-aaaa",
      "member": "false",
      "roles": [ "owner" ]
  }
}
```
On success, returns a `201 Created` response

Fixes #338

sbryzak commented:
Still a lot more implementation to do, plus testing.  Just wanted to push what I've done so far for review as we're up to 750 lines already and just really getting started on this PR. ;)

sbose78 commented on account/identity.go:
```diff
@@ -312,6 +329,13 @@ func IdentityWithUser() func(db *gorm.DB) *gorm.DB {
+// IdentithWithResource is a gorm filter for preloading the IdentityResource relationship
+func IdentityWithResource() func(db *gorm.DB) *gorm.DB {
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Preload("IdentityResource")
```
Need to check if this is a different query :)

sbryzak replied:
I've removed this function, gorm can't preload IdentityResource (and won't tell me why)

sbose78 commented on authorization/invitation/invitation.go:
```diff
@@ -0,0 +1,13 @@
+	uuid "github.com/satori/go.uuid"
+)
+
+type Invitation struct {
```
Nice, I've done something similar too here https://github.com/fabric8-services/fabric8-auth/pull/347/files#diff-bd9af08e13589624d0629a20880b3c70 because of the difference between the entity in the db, and the response json.

sbose78 commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,166 @@
+)
+
+type InvitationModelService interface {
+	CreateInvitations(ctx context.Context, issuingUserId uuid.UUID, inviteTo uuid.UUID, invitations []invitation.Invitation) error
```
Could just use `Create` and `List` ? to avoid stutter :) https://blog.golang.org/package-names

sbryzak replied:
Agreed, however I've renamed these methods differently now to account for different workflows.

sbose78 commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+// GormInvitationModelService is the implementation of the interface for
+// InvitationService. IMPORTANT NOTE: Transaction control is not provided by this service
```
nice note.

sbose78 commented on migration/sql-files/022-invitations.sql:
```diff
@@ -0,0 +1,35 @@
+create table invitation_role (
+  invitation_id uuid NOT NULL references invitation (invitation_id),
+  role_id uuid NOT NULL references role (role_id)
+  PRIMARY KEY (invitation_id, role_id)
```
Could we have an ID column which is an auto-generated uuid - and then make that the primary key.
We can make `(invitation_id, role_id)` a unique key.

Apart from general consistency with all other tables, this ensures that the database row cleanup logic doesn't break  https://github.com/fabric8-services/fabric8-auth/blob/master/gormsupport/cleaner/db_clean.go 
This helps the test data cleanup to happen 'automatically' using the primary key of the entity.

We don't need to use the ID anywhere in the code - apart from the repository struct.

sbose78 commented on design/invitation.go:
```diff
@@ -0,0 +1,58 @@
+var CreateInvitationRequestMedia = a.MediaType("application/vnd.create_invitation_request+json", func() {
+	a.Description("Request payload required to create new invitations")
+	a.Attributes(func() {
+		a.Attribute("invitees", a.ArrayOf(invitee), "An array of users invited to become members or to accept a role")
```
Could we please 'nest' it under `data` ? That'll help us add metadata in future without breaking any of the UI/clients.

sbryzak replied:
I've renamed `invitees` to `data`, will that work?

sbryzak commented:
Tests are looking pretty good now, I still need to add a few more (including controller tests) and we should be good for another review.  Will aim to get this completed by the end of the weekend

alexeykazakov commented on controller/invitation_blackbox_test.go:
```diff
@@ -0,0 +1,89 @@
+		},
+	}
+
+	test.CreateGroupInviteInvitationCreated(s.T(), s.Ctx, service, controller, orgIdentity.ID.String(), payload)
```
Use service.Ctx instead of s.Ctx

sbryzak commented:
[test]

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-355-23`

alexeykazakov commented on design/invitation.go:
```diff
@@ -0,0 +1,59 @@
+	a.Action("createGroupInvite", func() {
+		a.Security("jwt")
+		a.Routing(
+			a.POST("group/:inviteTo"),
```
Why POST `/invitations/group/:id` ? Why not just POST `/invitations/:id`?
/group/ is a bit confusing here IMO.

sbryzak replied:
URL has been updated to support both invitation types now.

alexeykazakov commented on design/invitation.go:
```diff
@@ -0,0 +1,59 @@
+		)
+		a.Description("Lists invitations for an organization, team or group that the user has the invite_user scope for, or lists invitations for the current user")
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		//a.Response(d.OK, invitationArray)
```
Let's remove this from now if we are not implementing it in this PR. Or at least return NotImplemented() for now.

sbryzak replied:
controller returns NotImplemented() now

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-355-24`

alexeykazakov commented on authorization/invitation/repository/invitation.go:
```diff
@@ -0,0 +1,243 @@
+	gormsupport.Lifecycle
+
+	// This is the primary key value
+	InvitationID uuid.UUID `sql:"type:uuid default uuid_generate_v4()" gorm:"primary_key" gorm:"column:invitation_id"`
```
`gorm:"primary_key;column:invitation_id"`

sbryzak replied:
Fixed

alexeykazakov commented on authorization/invitation/repository/invitation.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+type InvitationRole struct {
+	InvitationID uuid.UUID `sql:"type:uuid" gorm:"primary_key" gorm:"column:invitation_id"`
```
`gorm:"primary_key;column:invitation_id"`

sbryzak replied:
Fixed

alexeykazakov commented on authorization/invitation/repository/invitation.go:
```diff
@@ -0,0 +1,243 @@
+	InvitationID uuid.UUID `sql:"type:uuid" gorm:"primary_key" gorm:"column:invitation_id"`
+
+	Role   rolerepo.Role `gorm:"ForeignKey:RoleID;AssociationForeignKey:RoleID"`
+	RoleID uuid.UUID     `sql:"type:uuid" gorm:"primary_key" gorm:"column:role_id"`
```
`gorm:"primary_key;column:role_id"`

sbryzak replied:
Fixed

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,47 @@
+		return err
+	})
+
+	return err
```
Add //TODO send emails with a link to accept the invitations

sbryzak replied:
Done.

alexeykazakov commented on controller/invitation.go:
```diff
@@ -0,0 +1,82 @@
+
+// Create runs the create action.
+func (c *InvitationController) CreateInvite(ctx *app.CreateInviteInvitationContext) error {
+	currentUser, err := login.ContextIdentity(ctx)
```
Use login.ContextIdentityAndUserIfNotDeprovisioned()
and call it currentIdentity instead.

sbryzak replied:
I'm guessing you meant `LoadContextIdentityIfNotDeprovisioned()` - I've updated to use this function.

alexeykazakov commented on controller/invitation.go:
```diff
@@ -0,0 +1,82 @@
+		return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError(err.Error()))
+	}
+
+	if currentUser == nil {
```
currentUser/Identity can't be nil here

sbryzak replied:
Removed this check

alexeykazakov commented on design/invitation.go:
```diff
@@ -0,0 +1,59 @@
+	a.Attribute("username", d.String, "username of the user")
+	a.Attribute("user-email", d.String, "e-mail address of the user")
+	a.Attribute("member", d.Boolean, "if true invites the user to become a member")
+	a.Attribute("roles", a.ArrayOf(d.String), "An array of role names")
```
roles should be required, right?

sbryzak replied:
No, roles aren't required to be assigned, the invitation can be for membership only.

alexeykazakov commented on migration/sql-files/027-invitations.sql:
```diff
@@ -0,0 +1,22 @@
+);
+
+INSERT INTO resource_type_scope (resource_type_scope_id, resource_type_id, name) SELECT '87c2a3b2-6b7c-4d67-be4d-c73d5f51864b', resource_type_id, 'invite_user' FROM resource_type WHERE name = 'identity/organization';
+INSERT INTO role_scope (scope_id, role_id) SELECT '87c2a3b2-6b7c-4d67-be4d-c73d5f51864b', r.role_id FROM role r, resource_type rt WHERE r.resource_type_id = rt.resource_type_id AND r.name = 'owner' AND rt.name = 'identity/organization';
```
We need to keep all scope/role names in some common place so it's easy to track them.
Maybe in role package.

sbryzak replied:
They should be in `authorization/authorization.go`, that's where I've been defining the constants.

sbryzak commented on authorization/authorization.go:
```diff
@@ -12,4 +12,10 @@ const (
+	OwnerRole = "owner"
+
+	// InviteUserScope is the scope required for users wishing to invite other users to an organization, team or security group
+	InviteUserScope = "invite_user"
```
I've renamed this scope to "manage_members"

sbryzak commented on authorization/invitation/invitation.go:
```diff
@@ -0,0 +1,23 @@
+	Member     bool
+}
+
+/* This struct is used to provide a user with details of the invitations they have received.
```
Comments have been updated.

sbryzak commented on authorization/invitation/invitation.go:
```diff
@@ -0,0 +1,23 @@
+}
+
+/* This struct is used to provide a user with details of the invitations they have received.
+ * The ResourceType property indicates
```
Removed this line.

sbryzak commented on authorization/invitation/invitation.go:
```diff
@@ -0,0 +1,23 @@
+type InvitationDetail struct {
+	ResourceType string
+	Description  string
+	Member       bool
```
Descriptions added for all properties.

sbryzak commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,221 @@
+
+type InvitationModelService interface {
+	Issue(ctx context.Context, issuingUserId uuid.UUID, inviteTo string, invitations []invitation.Invitation) error
+	List(ctx context.Context, id uuid.UUID) ([]invitation.Invitation, error)
```
I'm not sure how, or even if we will be using these methods yet based on our discussion of simply sending an invitation e-mail for now and not having a dedicated UI screen to display a user's outstanding invitations.  Based on this, I've removed both the List and ListForUser methods for now.

sbryzak commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,221 @@
+type InvitationModelService interface {
+	Issue(ctx context.Context, issuingUserId uuid.UUID, inviteTo string, invitations []invitation.Invitation) error
+	List(ctx context.Context, id uuid.UUID) ([]invitation.Invitation, error)
+	ListForUser(ctx context.Context, id uuid.UUID) ([]invitation.InvitationDetail, error)
```
Method has been removed

sbryzak commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,221 @@
+	}
+}
+
+// Creates new invitations.  The inviteTo parameter is the unique id of the organization, team or security group for
```
Fixed the comment and added more detail.

sbryzak commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,221 @@
+				return errors.NewBadParameterErrorFromString("Username", invitation.UserName, "username not found")
+			} else if len(identities) > 1 {
+				// If more than one user is found, return an error
+				return errors.NewBadParameterErrorFromString("Username", invitation.UserName, "more than one user with same username found")
```
I've removed the len() > 1 check from here, let's discuss the other point separately.

sbryzak commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,221 @@
+		}
+
+		// Confirm that any specified roles are valid for this resource type
+		for _, roleName := range invitation.Roles {
```
I've removed this section.  The point of it was to validate the role names before we started creating records, but I guess that since this entire thing is in a transaction nothing will get created if the transaction fails.

sbryzak commented:
So there seems to be an issue here... after removing the role validation check the controller test is now failing at line 207:

`	require.Equal(s.T(), 0, len(invitations))`

This can only be due to invitation records actually being created, however the operation should have been cancelled and rolled back because the service returned an error, is my understanding correct here?

I can confirm that the error was returned because it's in the log:

`time="2018-04-14T12:46:32+07:00" level=error msg="failed to create invitations" err="Bad value for parameter 'Roles': 'foobar' - no such role found for resource type identity/organization" file="/home/shane/go/src/github.com/fabric8-services/fabric8-auth/controller/invitation.go" func="github.com/fabric8-services/fabric8-auth/controller.(*InvitationController).CreateInvite" identity_id=01e97fd3-dd09-4c65-8887-133d843cbe83 line=62 pid=15988 pkg=controller req_id= req_params=map[inviteTo:[05566841-2dcc-44ee-84ee-50a06c3eece2]] 
`

sbryzak commented:
Ok I found one issue, the controller test was creating the wrong service type (it was creating the model, not the service instance).  I've fixed that, and confirmed with the debugger that it is now calling the transaction-controlled service, however the transaction is still not being rolled back.  I've checked that the service method inside the transaction block is returning an error, and confirmed that it is, however I can query the database from the Postgres sql client and see that the record has still been created inside the table.  @alexeykazakov any ideas what might be happening here?

alexeykazakov commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,202 @@
+				return errors.NewBadParameterErrorFromString("Roles", roleName, fmt.Sprintf("no such role found for resource type %s", resourceTypeName))
+			}
+
+			s.repo.InvitationRepository().AddRole(ctx, inv.InvitationID, role.RoleID)
```
Check if it returns any error

alexeykazakov commented:
@sbryzak the bug was in the way you use repos inside a transaction. See the fix - https://github.com/fabric8-services/fabric8-auth/pull/355/commits/01befa326a8965d4acdd0cf03ed9188199ac3bbe

It's required to use the same repo (from appl application.Application) which is passed to the transaction function. We should clean up other model services and repos too.

alexeykazakov commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,201 @@
+			if !identity.IsUser() {
+				return errors.NewBadParameterErrorFromString("Identity ID", invitation.IdentityID, "identity is not a user")
+			}
+		} else if invitation.UserName != nil {
```
From my previous review:
> We should not do any look up by usernames or emails here. Our invitation API should accept identity(user) IDs only. In UI we will search for user anyway. So UI will know the identity ID. We don't have to do this search twice.

So, why do you think our API should support invite by usernames or email? Why not Identity ID only?

sbryzak replied:
After our discussion I've removed support for username and e-mail.

alexeykazakov commented on controller/invitation.go:
```diff
@@ -0,0 +1,80 @@
+func (c *InvitationController) CreateInvite(ctx *app.CreateInviteInvitationContext) error {
+	currentIdentity, err := login.LoadContextIdentityIfNotDeprovisioned(ctx, c.db)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError(err.Error()))
```
Don't need to wrap the error to errors.NewUnauthorizedError(err.Error())). 
login.LoadContextIdentityIfNotDeprovisioned(ctx, c.db) returns UnauthorizedError.
So, return jsonapi.JSONErrorResponse(ctx, err) is enough here.

sbryzak replied:
Fixed

alexeykazakov commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,201 @@
+			if len(users) == 0 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "user with e-mail address not found")
+			} else if len(users) > 1 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "more than one user with e-mail address found")
```
It can't happen. User emails are unique in the user table.

sbryzak replied:
This section has been removed

alexeykazakov commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,201 @@
+			if len(identities) == 0 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "no identity found for user with e-mail address")
+			} else if len(identities) > 1 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "more than one identity found for user with e-mail address")
```
It's not an error. We support multiple Identities per user. We don't use it right now though. This is for future use of remote identities (from GitHub, JIRA and other issue trackers).
You need to look up for identities with account.KeycloakIDP identity provider. Identities with that provider come from Keycloak/RHD and they represent OSIO users. There is supposed to be only one identity with such provider per user.
Feel free to add missing methods to the identity repo for that ;) But again, I don't think we should have look up for useremail at all in this API in the first place.

alexeykazakov commented on authorization/invitation/model/invitation_model_service_blackbox_test.go:
```diff
@@ -0,0 +1,256 @@
+	require.NoError(s.T(), err, "Error listing invitations")
+
+	require.Equal(s.T(), 1, len(invs))
+	require.True(s.T(), invs[0].Member)
```
Check all the relevant fields of the created invitation.

alexeykazakov commented on authorization/invitation/model/invitation_model_service_blackbox_test.go:
```diff
@@ -0,0 +1,256 @@
+
+	require.Equal(s.T(), 1, len(invs))
+	require.Equal(s.T(), invitee.ID, invs[0].IdentityID)
+	require.True(s.T(), invs[0].Member)
```
Check all the relevant fields of the created invitation.

alexeykazakov commented on authorization/invitation/model/invitation_model_service_blackbox_test.go:
```diff
@@ -0,0 +1,256 @@
+
+	require.Equal(s.T(), 1, len(invs))
+	require.Equal(s.T(), invitee.ID, invs[0].IdentityID)
+	require.True(s.T(), invs[0].Member)
```
Check all the relevant fields of the created invitation.

alexeykazakov commented on authorization/invitation/model/invitation_model_service_blackbox_test.go:
```diff
@@ -0,0 +1,256 @@
+	for _, inv := range invs {
+		if inv.IdentityID == invitee.ID {
+			found = true
+			require.True(s.T(), inv.Member)
```
Check all the relevant fields of the created invitation.

I think it's worth creating a function in this test which will assert the created invitation and use it all the relevant test functions.

alexeykazakov commented on authorization/invitation/repository/invitation_blackbox_test.go:
```diff
@@ -0,0 +1,131 @@
+	err = s.repo.Save(s.Ctx, &invitation)
+	require.NoError(s.T(), err)
+
+	require.True(s.T(), invitation.Member)
```
Check all the relevant field of the saved invitation.

sbryzak replied:
Done

alexeykazakov commented on authorization/invitation/repository/invitation_blackbox_test.go:
```diff
@@ -0,0 +1,131 @@
+	require.NoError(s.T(), err)
+
+	err = s.repo.AddRole(s.Ctx, invitation.InvitationID, role.RoleID)
+	require.NoError(s.T(), err)
```
Check if the role was added or not.

sbryzak replied:
Fixed

alexeykazakov commented on authorization/invitation/repository/invitation_blackbox_test.go:
```diff
@@ -0,0 +1,131 @@
+	require.NoError(s.T(), err)
+
+	require.Equal(s.T(), 1, len(roles))
+	require.Equal(s.T(), authorization.OwnerRole, roles[0].Name)
```
This test also tests AddRole(). It's OK to merge it to TestAddRole() test and have just one test TestAddAndListRoles() function which will cover both methods.

sbryzak replied:
Done

sbryzak commented on authorization/invitation/invitation.go:
```diff
@@ -0,0 +1,26 @@
+	// ResourceType is the name of the resource type to which the user has been issued an invitation
+	ResourceType string
+	// Description is a description of the organization, team, security group or resource to which the user has been issued an invitation
+	Description string
```
This struct is not actually required now since we will be sending an invitation e-mail, not displaying the invitations via UI screen.

sbryzak commented on authorization/invitation/model/invitation_model_service.go:
```diff
@@ -0,0 +1,201 @@
+	// If we get an error here, the value is definitely not for an Identity so we'll treat it as a resource ID
+	if err != nil {
+		// Try to lookup a resource with the same ID value
+		inviteToResource, err = repo.ResourceRepository().Load(ctx, inviteTo)
```
Well, since the Resource ID is the primary key for the RESOURCE table, we already have a unique constraint the same as every other table's primary key.  Meaning that attempting to create a resource with an ID that already exists will fail with an error.

sbryzak commented:
@alexeykazakov I've cleaned up the tests and hopefully improved coverage quite a bit.  I've refactored the service in line with the changes we are making in #446, we can use the other issue to address naming once we have completed our discussions on it.

sbryzak commented:
[test]

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-355-28`

sbose78 commented on migration/migration_blackbox_test.go:
```diff
@@ -299,6 +301,56 @@ func testMigration25ValidMiss(t *testing.T) {
+	require.NotNil(t, err)
+
+	// Cleanup the test data
+	require.Nil(t, runSQLscript(sqlDB, "026-cleanup-test-invitation-data.sql"))
```
optional, but good practice - since we anyway drop this db where we test migrations. ( after this test suite is done executing )

sbryzak commented on account/identity.go:
```diff
@@ -97,6 +97,10 @@ func (m Identity) GetLastModified() time.Time {
 	return m.UpdatedAt
 }
 
+func (m Identity) IsUser() bool {
```
Done.

sbryzak commented on authorization/authorization.go:
```diff
@@ -12,4 +12,16 @@ const (
+	ManageMembersScope = "manage_members"
 )
+
+func CanHaveMembers(resourceTypeName string) bool {
```
You won't believe this - this method had a bug which I only discovered when I wrote the test for it. ;)

sbryzak commented on authorization/invitation/repository/invitation.go:
```diff
@@ -0,0 +1,243 @@
+	return m.UpdatedAt
+}
+
+type InvitationRole struct {
```
Done

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-355-29`

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,154 @@
+		// Load the resource for the identity
+		identityResource, err = s.repo.ResourceRepository().Load(ctx, *inviteToIdentity.IdentityResourceID)
+		if err != nil {
+			return errors.NewInternalErrorFromString(ctx, "error loading resource for identity")
```
return errors.NewInternalError(ctx, err)

sbryzak replied:
Done.

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,154 @@
+		}
+
+		if !scope {
+			return errors.NewForbiddenError(fmt.Sprintf("user requires %s scope to invite other users", authorization.ManageMembersScope))
```
This is very important use case (when the user don't have permissions to invite) but it's not covered by tests - https://codecov.io/gh/fabric8-services/fabric8-auth/pull/355/diff#D28-79

sbryzak replied:
Test added.

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,154 @@
+
+		// We only allow membership in some identity types - confirm that we are inviting to an organization, team or security group
+		if !authorization.CanHaveMembers(identityResource.ResourceType.Name) {
+			return errors.NewInternalErrorFromString(ctx, "may only invite a user to an organization, team or security group")
```
Not covered by tests.

sbryzak replied:
Test added

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,154 @@
+			return errors.NewInternalErrorFromString(ctx, "may only invite a user to an organization, team or security group")
+		}
+	} else if inviteToResource != nil {
+		// Confirm that the issuing user has the manage members scope for the resource
```
This whole case when we invite to a resource is not covered by tests.
There should be two tests for this:
1. User has permissions to invite to a resource
2. User doesn't have permissions.

sbryzak replied:
I've added tests to cover both of these

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,154 @@
+		// Load the identity
+		identity, err := s.repo.Identities().Load(ctx, *invitation.IdentityID)
+		if err != nil {
+			return errors.NewInternalError(ctx, err)
```
No tests for this usecase (when we invite unknown user)

sbryzak replied:
Test added

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,154 @@
+		}
+
+		if !identity.IsUser() {
+			return errors.NewBadParameterErrorFromString("Identity ID", invitation.IdentityID, "identity is not a user")
```
No tests for this usecase (when the identity is not a user)

sbryzak replied:
Test added

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	// Iterate through all of the invitations and confirm that for each one:
+	// 1) a valid user has been specified via its User ID, e-mail address or username
```
Fix the comment. We now use user identity ID only.

sbryzak replied:
Comment has been fixed

alexeykazakov commented on authorization/invitation/repository/invitation.go:
```diff
@@ -0,0 +1,244 @@
+
+	err := m.db.CommonDB().QueryRow(query, id).Scan(&exists)
+	if err == nil && !exists {
+		return exists, errors.NewNotFoundError(m.TableName(), id.String())
```
Not covered by tests.

sbryzak replied:
Test added

alexeykazakov commented on authorization/invitation/service/invitation_service.go:
```diff
@@ -14,24 +20,135 @@ type InvitationService interface {
+// This method creates one record in the INVITATION table for each user in the invitations parameter.  Any roles that are issued
+// as part of a user's invitation are created in the INVITATION_ROLE table.
 func (s *InvitationServiceImpl) Issue(ctx context.Context, issuingUserId uuid.UUID, inviteTo string, invitations []invitation.Invitation) error {
+	var inviteToIdentity *account.Identity
```
You removed Model Service (need a better name) and moved its code (as-is) to the Controller Service.
This is not what we agreed to do yesterday. We can remove unnecessary layers but we can't mix them.
Controller Service must use transactions. But this service doesn't use them.
So I see two options here:
1. Move this code back to Model Service (need better name). Remove the Controller Service. Use Model Service directly from Controller. Transaction management should be in the controller then.
2. Wrap the whole thing inside this Issue() method in the Controller Service with a transaction. Remove transaction management from Controller.

IMO the first option is better.

alexeykazakov commented on authorization/permission/service/permission_service.go:
```diff
@@ -31,7 +32,7 @@ func NewPermissionModelService(db *gorm.DB) PermissionModelService {
 // certain type of resource in the resource ancestry to map to a role for a different resource type lower in the
 // resource hierarchy.
-func (s *GormPermissionModelService) HasScope(ctx context.Context, identityID uuid.UUID, resourceID string, scopeName string) (bool, error) {
+func (s *PermissionServiceImpl) HasScope(ctx context.Context, identityID uuid.UUID, resourceID string, scopeName string) (bool, error) {
```
Same thing. You are mixing Model Service (need a better name) with Controller Service.
You made this code a controller service. But then it must use transactions.
Just move this code back to the Model Service where it was before :)

alexeykazakov commented on controller/invitation.go:
```diff
@@ -16,47 +15,47 @@ import (
-			IdentityID: identityID,
-			UserName:   invitee.Username,
-			UserEmail:  invitee.UserEmail,
+			IdentityID: &identityID,
```
IdentityID here is actually required, right? So, we can make it UUID (not *UUID).

alexeykazakov commented on gormapplication/application.go:
```diff
@@ -97,6 +99,14 @@ func (g *GormBase) InvitationRepository() invitation.InvitationRepository {
+
+func (g *GormBase) PermissionService() permissionservice.PermissionService {
+	return permissionservice.NewPermissionService(g.db)
+}
```
Both method should return Model Services (need a better name). Not Controller Services.
Model Services is purely a DB layer.
Controller Services is not. They may use DB layer (if they do they should provide transaction control). They may call remote services.

sbryzak replied:
I've renamed these for now to put "Model" back into the name.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-355-30`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-355-31`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-355-32`
